parser-typechecker/src/Unison/Server/Endpoints/Projects.hs

```diff
@@ -0,0 +1,133 @@
+
+type ProjectsAPI =
+  "projects" :> QueryParam "rootBranch" ShortBranchHash
+    :> APIGet [ProjectListing]
+
+instance ToSample ProjectListing where
+  toSamples _ =
+    [ ( "Projects in the root branch",
+        ProjectListing
+          (ProjectOwner "unison")
+          "base"
+          "#gjlk0dna8dongct6lsd19d1o9hi5n642t8jttga5e81e91fviqjdffem0tlddj7ahodjo5"
+      )
+    ]
+
+data ProjectOwner = ProjectOwner Text deriving (Generic, Show)
+
+instance ToJSON ProjectOwner where
+  toEncoding = genericToEncoding defaultOptions
+
+deriving instance ToSchema ProjectOwner
+
+data ProjectListing = ProjectListing
+  { owner :: ProjectOwner,
+    name :: Text,
+    hash :: UnisonHash
+  }
+  deriving (Generic, Show)
+
+instance ToJSON ProjectListing where
+  toEncoding = genericToEncoding defaultOptions
+
+deriving instance ToSchema ProjectListing
+
+backendListEntryToProjectListing ::
+  Var v =>
+  ProjectOwner ->
+  Backend.ShallowListEntry v a ->
+  Maybe ProjectListing
+backendListEntryToProjectListing owner = \case
+  Backend.ShallowBranchEntry name hash _ ->
+    Just $
+      ProjectListing
+        { owner = owner,
+          name = NameSegment.toText name,
+          hash = "#" <> SBH.toText hash
+        }
+  _ -> Nothing
+
+entryToOwner ::
+  Var v =>
+  Backend.ShallowListEntry v a ->
+  Maybe ProjectOwner
+entryToOwner = \case
+  Backend.ShallowBranchEntry name _ _ ->
+    Just $ ProjectOwner $ NameSegment.toText name
+  _ -> Nothing
+
+serve ::
+  Var v =>
+  Handler () ->
+  Codebase IO v Ann ->
+  Maybe ShortBranchHash ->
+  Handler (APIHeaders [ProjectListing])
+serve tryAuth codebase mayRoot =
+  let errFromEither f = either (throwError . f) pure
+
+      doBackend a = do
+        ea <- liftIO $ runExceptT a
+        errFromEither backendError ea
+
+      findShallow branch = doBackend $ Backend.findShallowInBranch codebase branch
+
+      parsePath p = errFromEither (`badNamespace` p) $ Path.parsePath' p
+
+      ownerToProjectListings root owner = do
+        let (ProjectOwner ownerName) = owner
+        ownerPath' <- (parsePath . Text.unpack) ownerName
+        let path = Path.fromPath' ownerPath'
+        let ownerBranch = Branch.getAt' path root
+        entries <- findShallow ownerBranch
+        pure $ mapMaybe (backendListEntryToProjectListing owner) entries
+
+      projects = do
+        root <- case mayRoot of
+          Nothing -> do
+            gotRoot <- liftIO $ Codebase.getRootBranch codebase
+            errFromEither rootBranchError gotRoot
+          Just sbh -> do
+            ea <- liftIO . runExceptT $ do
+              h <- Backend.expandShortBranchHash codebase sbh
+              mayBranch <- lift $ Codebase.getBranchForHash codebase h
+              mayBranch ?? Backend.CouldntLoadBranch h
+            errFromEither backendError ea
+
+        ownerEntries <- findShallow root
+        let owners = mapMaybe entryToOwner ownerEntries
+        fmap join (traverse (ownerToProjectListings root) owners)
```

Comment:
I'm sure this whole function could be a lot cleaner!

Reply:
Looks like most of these definitions don't depend on closures, so would likely be nice to move them into a `where` to make the "main" logic a bit clearer, then maybe add some type-signatures too 😄 

Another reason to keep things in `where` is that oftentimes future changes to functions affect nesting, which means any `let` definitions need to be re-indented, and this causes trivial merge conflicts with any changes in any of those definitions even if one side just indented them 😬

Reply:
I think type annotations on the helper functions would help a lot (at least during the review process it certainly would :) )

But also, you might like `foldMapM` from `import Unison.Util.Monoid (foldMapM)` to help out here:

```suggestion
        foldMapM (ownerToProjectListings root) owners
```

Reply:
🤯

Reply:
Hmm, where would `foldMapM` come from?